src/color.rs

```diff
@@ -1289,3 +1287,61 @@ where
         alpha,
     ))
 }
+
+/// Try to parse the components and alpha with the legacy syntax, but also allow
+/// the [color-4] syntax if that fails.
+/// https://drafts.csswg.org/css-color-4/#color-syntax-legacy
+pub fn parse_legacy_components<'i, 't, P, F1, F2, F3, R1, R2, R3>(
+    color_parser: &P,
+    input: &mut Parser<'i, 't>,
+    f1: F1,
+    f2: F2,
+    f3: F3,
+) -> Result<(R1, R2, R3, f32), ParseError<'i, P::Error>>
+where
+    P: ColorParser<'i>,
+    F1: FnOnce(&P, &mut Parser<'i, 't>) -> Result<R1, ParseError<'i, P::Error>>,
+    F2: FnOnce(&P, &mut Parser<'i, 't>) -> Result<R2, ParseError<'i, P::Error>>,
+    F3: FnOnce(&P, &mut Parser<'i, 't>) -> Result<R3, ParseError<'i, P::Error>>,
+{
+    let r1 = f1(color_parser, input)?;
+    // Commas denote that we are using the legacy syntax.
+    let uses_commas = input.try_parse(|i| i.expect_comma()).is_ok();
+    let r2 = f2(color_parser, input)?;
+    if uses_commas {
+        input.expect_comma()?;
+    }
+    let r3 = f3(color_parser, input)?;
+
+    let alpha = parse_legacy_alpha(color_parser, input, uses_commas)?;
+
+    Ok((r1, r2, r3, alpha))
+}
+
+/// Parse the color components and alpha with the [color-4] syntax.
+pub fn parse_components<'i, 't, P, F1, F2, F3, R1, R2, R3>(
+    color_parser: &P,
+    input: &mut Parser<'i, 't>,
+    f1: F1,
+    f2: F2,
+    f3: F3,
+) -> Result<(R1, R2, R3, f32), ParseError<'i, P::Error>>
+where
+    P: ColorParser<'i>,
+    F1: FnOnce(&P, &mut Parser<'i, 't>) -> Result<R1, ParseError<'i, P::Error>>,
+    F2: FnOnce(&P, &mut Parser<'i, 't>) -> Result<R2, ParseError<'i, P::Error>>,
+    F3: FnOnce(&P, &mut Parser<'i, 't>) -> Result<R3, ParseError<'i, P::Error>>,
+{
+    let r1 = f1(color_parser, input)?;
+    let r2 = f2(color_parser, input)?;
+    let r3 = f3(color_parser, input)?;
+
+    let alpha = if !input.is_exhausted() {
```

Comment:
Seems we could keep having a single `parse_alpha`, but pass `/* uses_commas = */ false` here?